### ISSUE TITLE
Scheduling coroutines on cache-line misses

### ISSUE THREAD

**Issue by unicomp21** [WebAssembly/design#1298]
The future might end up being more about leveraging cpu-prefect, but if we could make this work, would be pretty amazing.  So far, Asyncify in Emscripten and the C++-20 coroutine support which just became available in the clang 8 compiler.  Clearly, we'll need help from Intel and AMD on this one.

https://www.linkedin.com/pulse/dont-stall-multitask-georgios-psaropoulos/

**Comment by sbc100** [comment]
Surly only a pre-emptive scheduler (such as the linux kernel + pthreads) could ever hope to do something like this?  Aren't co-routines inherently cooperative (i.e. they only yield when they choose too)?

Also, as far I know there is no way to observe cache line misses.   Even observing page faults isn't something that can normally be done in userspace is it?

**Comment by unicomp21** [comment]
Absolutely, they are cooperative.  But if it were possible to "lock" a bit field in the L1 or L2 cache which reflects the status of a "range" of cache lines, kinda like we do w/ page tables, some pretty cool stuff would become possible wrt cooperatively scheduling cache line prefetch, etc. while also running coroutines.

**Comment by unicomp21** [comment]
Like I said, we would most likely need additional instructions (ie microcode changes) for this to become possible.

**Comment by unicomp21** [comment]
This might be a good feature for a startup like Nuvia, fabless chip designer.
